File: examples/minimart-go-nuxt/backend/internal/infrastructure/controller/product_controller.go
package controller

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"minimart/internal/application/usecase"
)

// ProductController serves GET /products.
type ProductController struct {
	products *usecase.ProductUsecase
}

// NewProductController returns a ProductController.
func NewProductController(products *usecase.ProductUsecase) *ProductController {
	return &ProductController{products: products}
}

// List returns the active catalogue. It is the one endpoint a visitor can read
// without a cookie.
func (ctl *ProductController) List(c *gin.Context) {
	list, err := ctl.products.List(c.Request.Context())
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, list)
}

File: examples/minimart-java-next/backend/src/main/java/minimart/infrastructure/web/ProductController.java
package minimart.infrastructure.web;

import minimart.application.dto.ProductList;
import minimart.application.usecase.ProductUsecase;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/** Serves GET /products. */
@RestController
@RequestMapping("/api/v1")
public class ProductController {

  private final ProductUsecase products;

  /** A controller over {@code products}. */
  public ProductController(ProductUsecase products) {
    this.products = products;
  }

  /** Returns the active catalogue. It is the one endpoint a visitor can read without a cookie. */
  @GetMapping("/products")
  public ProductList list() {
    return products.list();
  }
}
